framework/aster-frame/src/mm/page_table/boot_pt.rs

```diff
@@ -0,0 +1,133 @@
+/// frames that are set up by the firmware, loader or the setup code.
+pub struct BootPageTable<
+    E: PageTableEntryTrait = PageTableEntry,
+    C: PagingConstsTrait = PagingConsts,
+> {
+    root_pt: FrameNumber,
+    // The frames allocated for this page table are not tracked with
+    // metadata [`crate::mm::frame::meta`]. Here is a record of it
+    // for deallocation.
+    frames: Vec<FrameNumber>,
+    _pretend_to_use: core::marker::PhantomData<(E, C)>,
+}
+
+impl<E: PageTableEntryTrait, C: PagingConstsTrait> BootPageTable<E, C> {
+    /// Create a new boot page table from the current page table root physical address.
+    ///
+    /// The caller must ensure that the current page table may be set up by the firmware,
+    /// loader or the setup code.
+    pub unsafe fn from_current_pt() -> Self {
+        let root_paddr = crate::arch::mm::current_page_table_paddr();
+        Self {
+            root_pt: root_paddr / C::BASE_PAGE_SIZE,
+            frames: Vec::new(),
+            _pretend_to_use: core::marker::PhantomData,
+        }
+    }
+
+    /// Map a base page to a frame.
+    /// This function will panic if the page is already mapped.
+    pub fn map_base_page(&mut self, from: Vaddr, to: FrameNumber, prop: PageProperty) {
+        let mut pt = self.root_pt;
+        let mut level = C::NR_LEVELS;
+        // Walk to the last level of the page table.
+        while level > 1 {
+            let index = pte_index::<C>(from, level);
+            let pte_ptr = unsafe { (paddr_to_vaddr(pt * C::BASE_PAGE_SIZE) as *mut E).add(index) };
+            let pte = unsafe { pte_ptr.read() };
+            pt = if !pte.is_present() {
+                let frame = self.alloc_frame();
+                unsafe { pte_ptr.write(E::new_pt(frame * C::BASE_PAGE_SIZE)) };
+                frame
+            } else if pte.is_last(level) {
+                panic!("mapping an already mapped huge page in the boot page table");
+            } else {
+                pte.paddr() / C::BASE_PAGE_SIZE
+            };
+            level -= 1;
+        }
+        // Map the page in the last level page table.
+        let index = pte_index::<C>(from, 1);
+        let pte_ptr = unsafe { (paddr_to_vaddr(pt * C::BASE_PAGE_SIZE) as *mut E).add(index) };
+        let pte = unsafe { pte_ptr.read() };
+        if pte.is_present() {
+            panic!("mapping an already mapped page in the boot page table");
+        }
+        unsafe { pte_ptr.write(E::new_frame(to * C::BASE_PAGE_SIZE, 1, prop)) };
+    }
+
+    fn alloc_frame(&mut self) -> FrameNumber {
+        let frame = FRAME_ALLOCATOR.get().unwrap().lock().alloc(1).unwrap();
+        self.frames.push(frame);
+        // Zero it out.
+        let vaddr = paddr_to_vaddr(frame * PAGE_SIZE) as *mut u8;
+        unsafe { core::ptr::write_bytes(vaddr, 0, PAGE_SIZE) };
+        frame
+    }
+}
+
+impl<E: PageTableEntryTrait, C: PagingConstsTrait> Drop for BootPageTable<E, C> {
+    fn drop(&mut self) {
```

Comment:
It seems that the safety of dropping `BootPageTable` must be conditional.

Reply:
maybe not planned now.

The boot process is pending for an overhaul, as stated in the TODO.

Reply:
Here is a simple fix to make the API of `BootPageTable` sound.

```rust
impl<E: PageTableEntryTrait, C: PagingConstsTrait> BootPageTable<E, C> {
    /// Retire this boot-stage page table.
    ///
    /// Do not drop a boot-stage page table. Instead, retire it. 
    ///
    /// # Safety
    ///
    /// This method can only be called when this boot-stage page table is no longer in use,
    /// e.g., after the permanent kernel page table has been activated.
    pub unsafe fn retire(mut self) {
        // Manually free all heap and frame memory allocated by this object
        {
            let mut frames = core::mem::take(&mut self.frames);
            for frame in frames {
                FRAME_ALLOCATOR.get().unwrap().lock().dealloc(*frame, 1);
            }
            drop(frames);
        }
        // We do not want or need to trigger drop
        forget(self);
    }
}

impl<E: PageTableEntryTrait, C: PagingConstsTrait> Drop for BootPageTable<E, C> {
    fn drop(&mut self) {
        panic("do not drop a boot-stage page table; instead, retire it.");
    }
}
```

In `init_kernel_page_table`, use the new `retire` method instead.